pyo3-derive-backend/src/pyenum.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use crate::common::{impl_extractext, impl_methods_inventory, impl_proto_registry};
+use proc_macro2::TokenStream;
+use quote::quote;
+
+pub fn build_py_enum(enum_: &syn::ItemEnum) -> syn::Result<TokenStream> {
+    let mut variants = Vec::new();
+
+    for variant in enum_.variants.iter() {
+        if !variant.fields.is_empty() {
+            return Err(syn::Error::new_spanned(
+                variant,
+                "#[pyenum] only supports unit enums",
+            ));
+        }
+        if let Some((_, syn::Expr::Lit(lit))) = &variant.discriminant {
+            variants.push((variant.ident.clone(), lit.clone()))
+        } else {
+            return Err(syn::Error::new_spanned(
+                variant,
+                "#[pyenum] requires explicit discriminant (MyVal = 4)",
+            ));
+        }
+    }
+
+    impl_enum(&enum_.ident, variants)
+}
+
+fn impl_enum(
+    enum_: &syn::Ident,
+    variants: Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let enum_cls = impl_class(enum_, None)?;
+    let variant_names: Vec<syn::Ident> = variants
+        .iter()
+        .map(|(ident, _)| variant_enumname(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let variant_cls = variants
+        .iter()
+        .map(|(ident, _)| impl_class(ident, Some(enum_)))
+        .collect::<syn::Result<Vec<_>>>()?;
+    let variant_consts = variants
+        .iter()
+        .map(|(ident, _)| impl_const(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let to_py = impl_to_py(enum_, &variants)?;
+    let from_py = impl_from_py(enum_, &variants)?;
+
+    Ok(quote! {
+
+        #enum_cls
+
+        #(
+            struct #variant_names;
+        )*
+
+        #(
+            #variant_cls
+        )*
```

Comment:
I don't think this bit is needed - instead you should just be making the variants be instances of the the enum class. (Like how it is with the `enum` module.)